### ISSUE TITLE
[C#] Implement new Concept Exercise: resource cleanup

### ISSUE THREAD

**Issue by mikedamay** [exercism/v3#1639]
This issue describes how to implement the `resource-cleanup` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student how to use regular expressions in C#.

## Learning objectives

- Know how to clean up resources with `IDisposable` in C#.

## Out of scope

- resource lifetime - `using`

## Concepts

- `resource-cleanup` with `IDisposable in C# and .NET.

## Prerequisites

- `classes`
- `interfaces`

Other prerequisites will depend on the story.

### Hints
The [disposable pattern][disposable-patter] documentation details how `IDisposable` should be used.
The [finalizer][finalizers] documentation discusses the role of object finaliizers.

### After

The [disposable pattern][disposable-patter] documentation details how `IDisposable` should be used.
The [finalizer][finalizers] documentation discusses the role of object finaliizers.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[disposable-pattern]: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose
[finalizers]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/destructors
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer

**Comment by mikedamay** [comment]
I'm thinking about some sort of image/video display class with a method to display (acquire memory) and a method to close (requiring immediate disposal).  

Any  thoughts?

**Comment by ErikSchierboom** [comment]
`resource lifetime - using` is out of scope here. I can sort of understand why, but on the other hand `IDisposable` and `using` statement go hand in hand. Will this be just about the `IDisposable` interface?

> The finalizer documentation discusses the role of object finaliizers.

The hints mention finalizers. Will they be included in this exercise?

> I'm thinking about some sort of image/video display class with a method to display (acquire memory) and a method to close (requiring immediate disposal).

Would the memory acquisition be something like a `MemoryStream`? What would the student be expected to do with the memory?

**Comment by mikedamay** [comment]
> resource lifetime - using is out of scope here.

My thinking is that immediately you are aware of `using` you forget about `IDisposable` etc.  By having one exercise that addresses `IDisposable` that gets the point over and then a separate one that covers the two forms of `using` can build on this.  This separation will help the student understand the mechanisms underlying `using` which is probably a good thing.

> The hints mention finalizers. Will they be included in this exercise?

I was thinking the only place finalizers gets a look in is _after.md_.  I do not know a good use case (which also supports good practice) other than using them for diagnostics.  Perhaps a diagnostic requirement could be part of a story but it would be impossible to test but that is not the end of the world.  What do you think?

> Would the memory acquisition be something like a MemoryStream? 

Actually what I was thinking of wouldn't work.  I can't see how we can show this off without hinting at external resources.  Maybe a string reader (standing in for a stream reader - I've always assumed that `IDisposable.Dispose()` was otiose in the case of memory-based readers/streams) would be the easiest to handle and do something with.

**Comment by ErikSchierboom** [comment]
> My thinking is that immediately you are aware of using you forget about IDisposable etc. By having one exercise that addresses IDisposable that gets the point over and then a separate one that covers the two forms of using can build on this. This separation will help the student understand the mechanisms underlying using which is probably a good thing.

Excellent explanation. Makes total sense.

> I was thinking the only place finalizers gets a look in is after.md. I do not know a good use case (which also supports good practice) other than using them for diagnostics. Perhaps a diagnostic requirement could be part of a story but it would be impossible to test but that is not the end of the world. What do you think?

I like the `after.md` document, as I also have rarely (if ever) seen a good use-case for them.

> Actually what I was thinking of wouldn't work. I can't see how we can show this off without hinting at external resources. Maybe a string reader (standing in for a stream reader - I've always assumed that IDisposable.Dispose() was otiose in the case of memory-based readers/streams) would be the easiest to handle and do something with.

If we check [the docs](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose), it states that:

> Implementing the Dispose method is primarily for releasing unmanaged resources used by your code.

This is slightly problematic for us, as unmanaged resources is not something the student will know about. The only resource that I feel could make sense are (file) streams, but that's not optimal either.

The docs also state:

> There are additional reasons for implementing Dispose, such as undoing something that was previously done. For example, freeing memory that was allocated, removing an item from a collection that was added, signaling the release of a lock that was acquired, and so on.

This got me thinking a bit. What if we could have the student model a database transaction. It could have `Commit` and `Rollback` methods and by default, it would commit when being disposed of (it not already being committed).

P.S. I don't know what otiose means :)

**Comment by mikedamay** [comment]
> P.S. I don't know what otiose means :)

The first thing to know, if you ever use it, is to pronounce it oweshus (like ocean but ...shus).  The point I was attempting to make was that memory-based streams and readers do not use external resources so including the dispose pattern for them has no immediate purpose.

A database transaction simulation sounds promising.  I will sketch it out.

**Comment by ErikSchierboom** [comment]
> The first thing to know, if you ever use it, is to pronounce it oweshus (like ocean but ...shus)

Wow, I would never have guessed. Reminds me of how I completely fumbled the pronunciation of `quay` :)

**Comment by mikedamay** [comment]
@ErikSchierboom 

The student could be tasked to write an ORM comprising `OpenTransaction()` to get a transaction from the database, `WriteObject()` to write in the context of a transaction and `CommitTransaction()` to commit what has been written.

The exercise provides a database object with methods to `CreateTransaction()` to create a transaction,`Write()` in the context of the transaction and `Transaction.Commit()` to commit a transaction.  It also implements the `IDisposable` interface.

Task 1 - Implement `ORM.OpenTransaction()` - the tests ensure a call to `Database.CreateTransaction()` was made.

Task 2 - Implement `ORM.WriteObject()` - the tests ensure that a a request to `Database.Write()` is made passing in a transaction.  If `Database.Write()` fails then the transaction must be disposed of

Task 3 - Implement `ORM.CommitTransaction()` - the test ensure that a request to `Transaction.Commit()` is made.  If `Transaction.Commit()` throws an error then the transaction must be disposed of.

Task 4 - Implement `IDispoosable` on the ORM.

The semantics are not a million miles away from .NET DB providers like `OleDb`.  We could slightly simplify things by making the `WriteObject()` a call on the transaction rather than the database but maybe wait and see what it looks like.

What do you think?

**Comment by ErikSchierboom** [comment]
I like it. I was wondering if we need the `ORM` class and instead could just have the `Transaction` and `Database` classes? This makes it a little less realistic, but does simplify things a bit. The `IDisposable` implementation could then be on the `Transaction` class, which is also how Entity Framework does it I believe (see [these docs](https://docs.microsoft.com/en-us/ef/core/saving/transactions)).

**Comment by mikedamay** [comment]
I'm all for making it simpler (even if more artificial) but I'm struggling with your vision.

Is the student responsible for implementing the `Database` class as well as the `Transaction` class?  If so how do we cause a failure?

If not, where would the experience of handling an exception and disposing of `Transaction` in a `finally` clause or whatever come from?

**Comment by mikedamay** [comment]
I hope these [instructions](https://github.com/mikedamay/v3/blob/csharp/resource-cleanup/languages/csharp/exercises/concept/resource-cleanup/.docs/instructions.md) and this [code](https://github.com/mikedamay/v3/blob/csharp/resource-cleanup/languages/csharp/exercises/concept/resource-cleanup/.meta/Example.cs) can form the basis of discussion.  There are a number of competing challenges.

**Comment by ErikSchierboom** [comment]
I really like the instructions. They are very clear on what to do and the flow works nicely. The code for the `Orm` class is excellent. If we have students write that they will definitely have a nice grasp of how to work with `IDisposable`. The `Database` class is quite involved though and, if presented to the student, will give away part of the solution on how to implement `IDisposable`. Did you mean to use the `Database` class as some sort of reference for the student? One idea that just came up with me is that the `Database` class could also be an interface that we can mock in the test suite.

**Comment by mikedamay** [comment]
The idea was that we would provide the database implementation.  I agree that a disadvantage of this is that it gives away a lot.

We are trying to teach the student 2 things:
1. How to **use** a class that implements `IDisposable` i.e. ensure you have catches and finallys to handle problems.  This will lead into the `resource-lifetime` exercise with the `using` keyword.
2. How to **implement** a class with the dispose pattern - include a finalzer, include a protected member or whatever.

We also need to have access to the state of the object implementing `IDisposable` in this case `Database` so that we can cause it to break (throw an exception) and tell in tests that it is being used correctly.

Could we stick the database class in the test file?

We do address issue 2, implementation, but the `IDisposable` pattern becomes very trivial when dealing with managed resources and brings noise and complexity.

I like putting the database class in the test file.  I do worry about noise and complexity in the template code but am less concerned than you are about the student getting a glimpse of "how the sausage is made".

**Comment by ErikSchierboom** [comment]
> I favour putting the database class in the test file. I do worry about noise and complexity in the template code but am less concerned than you are about the student getting too many clues.

I'm fine with that. We do have to remember that CLI-based students will get to see that class, so anything we can do to make that class as simple as possible sounds like a good idea.

**Comment by mikedamay** [comment]
@ErikSchierboom 
> This is slightly problematic for us, as unmanaged resources is not something the student will know about. The only resource that I feel could make sense are (file) streams, but that's not optimal either.

I have had a scout around the BCL source and have gained some insight (or misunderstood) and I'm not sure whether or not this is news to you as well.  When the docs talk about unmanaged resources in the context of the `IDisposable` pattern they have in mind that the class that implements `IDisposable` and follows the pattern is the class that directly acquires the unmanaged resource, presumably by some `extern` call.  When we use `System.IO.Stream` from our point of view that is a managed resource and a simple call to `IDispose` is all that is required - no need to involve finalizers.

Assuming my interpretation of _unmanaged_ is correct there is no way that we can have the student be involved with `extern` calls.  I'm not even sure how that works outside of Win32.  Therefore, I don't really think it does any harm to show the student the `Database` class which will be faking the idea of an external resource and I can add calls to that effect or at least talk about them.  If they copy the `IDisposable` pattern from `Database` they will be getting it wrong and we can correct the misapprehension I've had for the last 20 years.

**Comment by ErikSchierboom** [comment]
> Assuming my interpretation of unmanaged is correct there is no way that we can have the student learning about P/Invoke, finalizers and the obscure dispose pattern. We can punt the dispose pattern to a very advanced exercise.

Excellent. Thanks for explaining this!

**Comment by mikedamay** [comment]
closed by #1778
